apps/etterlatte-brev-api/src/main/kotlin/no/nav/etterlatte/brev/vedtaksbrev/VedtaksbrevServiceNy.kt

```diff
@@ -0,0 +1,286 @@
+            return
+        } else if (!brev.kanEndres()) {
+            throw UgyldigStatusKanIkkeFerdigstilles(brev.id, brev.status)
+        } else if (brev.mottakere.size !in 1..2) {
+            logger.error("Brev ${brev.id} har ${brev.mottakere.size} mottakere. Dette skal ikke være mulig...")
+            throw UgyldigAntallMottakere()
+        } else if (brev.mottakere.any { it.erGyldig().isNotEmpty() }) {
+            sikkerlogger.error("Ugyldig mottaker(e): ${brev.mottakere.toJson()}")
+            throw UgyldigMottakerKanIkkeFerdigstilles(brev.id, brev.sakId, brev.mottakere.flatMap { it.erGyldig() })
+        }
+
+        logger.info("Ferdigstiller brev med id=${brev.id}")
+        if (db.hentPdf(brev.id) == null) {
+            throw BrevManglerPDF(brev.id)
+        } else {
+            db.settBrevFerdigstilt(brev.id, brukerTokenInfo)
+        }
+    }
+
+    suspend fun tilbakestillVedtaksbrev(
+        brevId: Long,
+        bruker: BrukerTokenInfo,
+        brevRequest: BrevRequest,
+    ): BrevService.BrevPayload {
+        val brev = db.hentBrev(brevId)
+        if (!brev.kanEndres()) {
+            throw UgyldigForespoerselException(
+                "BREVET_KAN_IKKE_ENDRES",
+                "Kan ikke oppdatere brevet med id=$brevId, fordi brevet har status (${brev.status})",
+            )
+        }
+
+        val (spraak, sak, innsender, soeker, avdoede, verge, saksbehandlerIdent, attestantIdent, _, brevInnholdData) = brevRequest
+
+        val brevKode = brevInnholdData.brevKode
+        val avsender = utledAvsender(bruker, saksbehandlerIdent, attestantIdent, sak.enhet)
+
+        // val spraak = db.hentBrevInnhold(brevId)?.spraak TODO ??
+
+        val brevinnhold =
+            BrevInnhold(
+                brevKode.titlerPaaSpraak[spraak] ?: brevKode.tittel,
+                spraak,
+                brevbaker.hentRedigerbarTekstFraBrevbakeren(
+                    BrevbakerRequest.fra(
+                        brevKode = brevKode.redigering,
+                        brevData = ManueltBrevData(),
+                        avsender = avsender,
+                        soekerOgEventuellVerge = SoekerOgEventuellVerge(soeker, verge),
+                        sakId = sak.id,
+                        spraak = spraak,
+                        sakType = sak.sakType,
+                    ),
+                ),
+            )
+
+        if (brevinnhold.payload != null) {
+            db.oppdaterPayload(brevId, brevinnhold.payload, bruker)
+        }
+
+        /* TODO vedlegg er ikke redigerbare for tilbakereving?
+        if (innholdVedlegg != null) {
+            db.oppdaterPayloadVedlegg(brevId, innholdVedlegg, bruker)
+        }
+
+        // TODO kan brevkode endre seg?
+        if (opprinneligBrevkoder != brevkode) {
+            db.oppdaterBrevkoder(brevId, brevkode)
+            db.oppdaterTittel(brevId, brevinnhold.tittel, bruker)
+        }
```

Comment:
For førstegangsbehandling / revurdering så kan brevkoder endre seg hvis vi enten endrer utfallet i vedtaket, eller går mellom foreldreløs / ikke foreldreløs i BP. Men det er sikkert verdt å se litt på de tilfellene når vi kommer til de brevene